File: apps/mobile/android/app/src/main/java/com/healthcoachai/app/services/CostOptimizationService.kt
package com.healthcoachai.app.services

import com.healthcoachai.app.data.*
import com.healthcoachai.app.network.ApiClient
import com.healthcoachai.app.network.ApiResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * AI Cost Optimization Service
 * Manages AI usage tracking, cost optimization, and quota monitoring
 */
class CostOptimizationService(private val apiClient: ApiClient) {

    private val _costMetrics = MutableStateFlow<CostMetrics?>(null)
    val costMetrics: StateFlow<CostMetrics?> = _costMetrics.asStateFlow()
    
    private val _quotaStatus = MutableStateFlow<QuotaStatus?>(null)
    val quotaStatus: StateFlow<QuotaStatus?> = _quotaStatus.asStateFlow()
    
    private val _isNearLimit = MutableStateFlow(false)
    val isNearLimit: StateFlow<Boolean> = _isNearLimit.asStateFlow()
    
    private val _isOverLimit = MutableStateFlow(false)
    val isOverLimit: StateFlow<Boolean> = _isOverLimit.asStateFlow()

    /**
     * Get cost metrics for current user
     */
    suspend fun getCostMetrics(userId: String): ApiResult<CostMetrics> {
        val result = apiClient.request<CostMetrics>(
            endpoint = "/ai-prompt-optimization/cost-metrics?userId=$userId",
            method = "GET"
        )
        
        if (result is ApiResult.Success) {
            _costMetrics.value = result.data
        }
        
        return result
    }

    /**
     * Get quota status for current user
     */
    suspend fun getQuotaStatus(userId: String): ApiResult<QuotaStatus> {
        val result = apiClient.request<QuotaStatus>(
            endpoint = "/ai-prompt-optimization/quota-status?userId=$userId",
            method = "GET"
        )
        
        if (result is ApiResult.Success) {
            _quotaStatus.value = result.data
            _isNearLimit.value = result.data.isNearLimit
            _isOverLimit.value = result.data.isOverLimit
        }
        
        return result
    }

    /**
     * Execute optimized prompt template
     */
    suspend fun executeOptimizedPrompt(
        userId: String,
        templateId: String,
        variables: Map<String, Any>,
        category: String = "general"
    ): ApiResult<OptimizedPromptResponse> {
        val requestBody = mapOf(
            "userId" to userId,
            "templateId" to templateId,
            "variables" to variables,
            "category" to category
        )
        
        val result = apiClient.request<OptimizedPromptResponse>(
            endpoint = "/ai-prompt-optimization/execute",
            method = "POST",
            body = requestBody
        )
        
        // Update quota status if provided
        if (result is ApiResult.Success) {
            result.data.quotaStatus?.let { quotaStatus ->
                _quotaStatus.value = quotaStatus
                _isNearLimit.value = quotaStatus.isNearLimit
                _isOverLimit.value = quotaStatus.isOverLimit
            }
        }
        
        return result
    }

    /**
     * Get available prompt templates
     */
    suspend fun getTemplates(category: String? = null): ApiResult<List<PromptTemplate>> {
        val endpoint = if (category != null) {
            "/ai-prompt-optimization/templates?category=$category"
        } else {
            "/ai-prompt-optimization/templates"
        }
        
        return apiClient.request(endpoint)
    }

    /**
     * Track API usage for cost optimization
     */
    suspend fun trackUsage(
        userId: String,
        endpoint: String,
        method: String,
        tokenCount: Int? = null,
        cost: Double? = null
    ): ApiResult<Unit> {
        val requestBody = mapOf(
            "userId" to userId,
            "endpoint" to endpoint,
            "method" to method,
            "tokenCount" to tokenCount,
            "cost" to cost,
            "timestamp" to System.currentTimeMillis()
        )
        
        return apiClient.request(
            endpoint = "/ai-prompt-optimization/track-usage",
            method = "POST",
            body = requestBody
        )
    }

    /**
     * Get cost optimization settings
     */
    suspend fun getSettings(userId: String): ApiResult<CostOptimizationSettings> {
        return apiClient.request("/ai-prompt-optimization/settings?userId=$userId")
    }

    /**
     * Update cost optimization settings
     */
    suspend fun updateSettings(
        userId: String,
        settings: CostOptimizationSettings
    ): ApiResult<Unit> {
        return apiClient.request(
            endpoint = "/ai-prompt-optimization/settings",
            method = "PUT",
            body = mapOf("userId" to userId, "settings" to settings)
        )
    }

    /**
     * Check if user is approaching cost limits
     */
    fun shouldOptimizeForCost(): Boolean {
        val quota = _quotaStatus.value
        return quota?.isNearLimit == true || quota?.isOverLimit == true
    }

    /**
     * Get recommended optimization actions
     */
    fun getOptimizationRecommendations(): List<String> {
        val recommendations = mutableListOf<String>()
        val quota = _quotaStatus.value
        val metrics = _costMetrics.value

        if (quota?.isOverLimit == true) {
            recommendations.add("You've exceeded your daily quota. Consider upgrading your plan.")
        } else if (quota?.isNearLimit == true) {
            recommendations.add("You're approaching your daily limit. Consider using optimized templates.")
        }

        metrics?.let {
            if (it.averageCostPerRequest > 0.05) {
                recommendations.add("Your average cost per request is high. Try using batch processing.")
            }
            
            if (it.averageTokensPerRequest > 1000) {
                recommendations.add("Consider using shorter, more focused prompts to reduce token usage.")
            }
        }

        return recommendations
    }
}

// MARK: - Data Models

@kotlinx.serialization.Serializable
data class CostMetrics(
    val totalRequests: Int,
    val totalTokens: Int,
    val totalCost: Double,
    val averageTokensPerRequest: Double,
    val averageCostPerRequest: Double,
    val costByModel: Map<String, Double>,
    val tokensByModel: Map<String, Int>,
    val requestsByCategory: Map<String, Int>,
    val dailyCost: Double,
    val monthlyCost: Double,
    val projectedMonthlyCost: Double
)

@kotlinx.serialization.Serializable
data class QuotaStatus(
    val userId: String,
    val dailyQuota: Int,
    val dailyUsed: Int,
    val monthlyQuota: Int,
    val monthlyUsed: Int,
    val isNearLimit: Boolean,
    val isOverLimit: Boolean,
    val resetTime: String // ISO 8601 timestamp
)

@kotlinx.serialization.Serializable
data class OptimizedPromptResponse(
    val success: Boolean,
    val response: String,
    val templateId: String,
    val tokenCount: Int,
    val costUsd: Double,
    val optimizationRatio: Double,
    val quotaStatus: QuotaStatus? = null,
    val executionTime: Double
)

@kotlinx.serialization.Serializable
data class PromptTemplate(
    val id: String,
    val name: String,
    val description: String,
    val category: String,
    val template: String,
    val variables: List<String>,
    val estimatedTokens: Int,
    val averageCost: Double,
    val usageCount: Int,
    val successRate: Double
)

@kotlinx.serialization.Serializable
data class CostOptimizationSettings(
    val enableBatching: Boolean,
    val maxBatchSize: Int,
    val batchTimeoutSeconds: Int,
    val enableTemplateOptimization: Boolean,
    val costThresholdUsd: Double,
    val preferredModel: String? = null
)

File: apps/mobile/ios/HealthCoachAI/Services/CostOptimizationService.swift
import Foundation
import Combine

// MARK: - AI Cost Optimization Service

class CostOptimizationService: ObservableObject {
    static let shared = CostOptimizationService()
    
    private let apiService = APIService.shared
    @Published var currentCostMetrics: CostMetrics?
    @Published var quotaStatus: QuotaStatus?
    @Published var isNearLimit = false
    @Published var isOverLimit = false
    
    private init() {
        startPeriodicUpdates()
    }
    
    // MARK: - Cost Tracking
    
    /**
     * Get cost metrics for current user
     */
    func getCostMetrics(userId: String) async throws -> CostMetrics {
        guard let metrics = try await apiService.request(
            endpoint: "/ai-prompt-optimization/cost-metrics?userId=\(userId)",
            method: .GET,
            responseType: CostMetrics.self
        ) else {
            throw APIError.decodingError(NSError(domain: "CostOptimizationService", code: 0, userInfo: [NSLocalizedDescriptionKey: "Failed to get cost metrics"]))
        }
        
        await MainActor.run {
            self.currentCostMetrics = metrics
        }
        
        return metrics
    }
    
    /**
     * Get quota status for current user
     */
    func getQuotaStatus(userId: String) async throws -> QuotaStatus {
        guard let status = try await apiService.request(
            endpoint: "/ai-prompt-optimization/quota-status?userId=\(userId)",
            method: .GET,
            responseType: QuotaStatus.self
        ) else {
            throw APIError.decodingError(NSError(domain: "CostOptimizationService", code: 0, userInfo: [NSLocalizedDescriptionKey: "Failed to get quota status"]))
        }
        
        await MainActor.run {
            self.quotaStatus = status
            self.isNearLimit = status.isNearLimit
            self.isOverLimit = status.isOverLimit
        }
        
        return status
    }
    
    /**
     * Execute optimized prompt template
     */
    func executeOptimizedPrompt(
        userId: String,
        templateId: String,
        variables: [String: Any],
        category: String = "general"
    ) async throws -> OptimizedPromptResponse {
        let requestBody = [
            "userId": userId,
            "templateId": templateId,
            "variables": variables,
            "category": category
        ] as [String: Any]
        
        guard let response = try await apiService.request(
            endpoint: "/ai-prompt-optimization/execute",
            method: .POST,
            body: requestBody,
            responseType: OptimizedPromptResponse.self
        ) else {
            throw APIError.decodingError(NSError(domain: "CostOptimizationService", code: 0, userInfo: [NSLocalizedDescriptionKey: "Failed to execute optimized prompt"]))
        }
        
        // Update quota status if provided
        if let quotaStatus = response.quotaStatus {
            await MainActor.run {
                self.quotaStatus = quotaStatus
                self.isNearLimit = quotaStatus.isNearLimit
                self.isOverLimit = quotaStatus.isOverLimit
            }
        }
        
        return response
    }
    
    /**
     * Get available prompt templates
     */
    func getTemplates(category: String? = nil) async throws -> [PromptTemplate] {
        var endpoint = "/ai-prompt-optimization/templates"
        if let category = category {
            endpoint += "?category=\(category)"
        }
        
        guard let templates = try await apiService.request(
            endpoint: endpoint,
            method: .GET,
            responseType: [PromptTemplate].self
        ) else {
            return []
        }
        
        return templates
    }
    
    /**
     * Track API usage for cost optimization
     */
    func trackUsage(
        userId: String,
        endpoint: String,
        method: String,
        tokenCount: Int?,
        cost: Double?
    ) async {
        let requestBody = [
            "userId": userId,
            "endpoint": endpoint,
            "method": method,
            "tokenCount": tokenCount as Any,
            "cost": cost as Any,
            "timestamp": ISO8601DateFormatter().string(from: Date())
        ] as [String: Any]
        
        do {
            _ = try await apiService.request(
                endpoint: "/ai-prompt-optimization/track-usage",
                method: .POST,
                body: requestBody,
                responseType: EmptyResponse.self
            )
        } catch {
            // Log error but don't throw - usage tracking shouldn't break main functionality
            print("Failed to track usage: \(error.localizedDescription)")
        }
    }
    
    /**
     * Get cost optimization settings
     */
    func getSettings(userId: String) async throws -> CostOptimizationSettings {
        guard let settings = try await apiService.request(
            endpoint: "/ai-prompt-optimization/settings?userId=\(userId)",
            method: .GET,
            responseType: CostOptimizationSettings.self
        ) else {
            // Return default settings if none found
            return CostOptimizationSettings(
                enableBatching: true,
                maxBatchSize: 5,
                batchTimeoutSeconds: 30,
                enableTemplateOptimization: true,
                costThresholdUsd: 10.0,
                preferredModel: nil
            )
        }
        
        return settings
    }
    
    /**
     * Update cost optimization settings
     */
    func updateSettings(userId: String, settings: CostOptimizationSettings) async throws {
        _ = try await apiService.request(
            endpoint: "/ai-prompt-optimization/settings",
            method: .PUT,
            body: ["userId": userId, "settings": settings],
            responseType: EmptyResponse.self
        )
    }
    
    // MARK: - Periodic Updates
    
    private func startPeriodicUpdates() {
        // Update cost metrics every 5 minutes
        Timer.scheduledTimer(withTimeInterval: 300, repeats: true) { [weak self] _ in
            Task {
                // In a real app, get the actual user ID from auth context
                let userId = "user_123"
                try? await self?.getCostMetrics(userId: userId)
                try? await self?.getQuotaStatus(userId: userId)
            }
        }
    }
    
    // MARK: - Utility Methods
    
    /**
     * Check if user is approaching cost limits
     */
    func shouldOptimizeForCost() -> Bool {
        guard let quotaStatus = quotaStatus else { return false }
        return quotaStatus.isNearLimit || quotaStatus.isOverLimit
    }
    
    /**
     * Get recommended optimization actions
     */
    func getOptimizationRecommendations() -> [String] {
        var recommendations: [String] = []
        
        if isOverLimit {
            recommendations.append("You've exceeded your daily quota. Consider upgrading your plan.")
        } else if isNearLimit {
            recommendations.append("You're approaching your daily limit. Consider using optimized templates.")
        }
        
        if let metrics = currentCostMetrics {
            if metrics.averageCostPerRequest > 0.05 {
                recommendations.append("Your average cost per request is high. Try using batch processing.")
            }
            
            if metrics.averageTokensPerRequest > 1000 {
                recommendations.append("Consider using shorter, more focused prompts to reduce token usage.")
            }
        }
        
        return recommendations
    }
}

// MARK: - Supporting Types

struct OptimizedPromptResponse: Codable {
    let success: Bool
    let response: String
    let templateId: String
    let tokenCount: Int
    let costUsd: Double
    let optimizationRatio: Double
    let quotaStatus: QuotaStatus?
    let executionTime: Double
}

struct PromptTemplate: Codable, Identifiable {
    let id: String
    let name: String
    let description: String
    let category: String
    let template: String
    let variables: [String]
    let estimatedTokens: Int
    let averageCost: Double
    let usageCount: Int
    let successRate: Double
}